File: rainbow/c/rainbow.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_led_strip.h"
#include "config.h"

uint8_t leds[LED_ROWS*LED_COLS][3] = {{0}};
uint8_t rainbow[LED_ROWS*LED_COLS][3] = {{0}};
float rainbow_position = 0;

// based on http://web.mit.edu/storborg/Public/hsvtorgb.c
void hsv_to_rgb(uint8_t h, uint8_t s, uint8_t v, uint8_t *r, uint8_t *g, uint8_t *b) {
	uint8_t region, fpart, p, q, t;

	if(s == 0) {
		// color is grayscale
		*r = *g = *b = v;
		return;
	}

	// make hue 0-5
	region = h / 43;

	// find remainder part, make it from 0-255
	fpart = (h - (region * 43)) * 6;

	// calculate temp vars, doing integer multiplication
	p = (v * (255 - s)) >> 8;
	q = (v * (255 - ((s * fpart) >> 8))) >> 8;
	t = (v * (255 - ((s * (255 - fpart)) >> 8))) >> 8;

	// assign temp vars based on color cone region
	switch(region) {
	case 0:  *r = v; *g = t; *b = p; break;
	case 1:  *r = q; *g = v; *b = p; break;
	case 2:  *r = p; *g = v; *b = t; break;
	case 3:  *r = p; *g = q; *b = v; break;
	case 4:  *r = t; *g = p; *b = v; break;
	default: *r = v; *g = p; *b = q; break;
	}
}

void frame_upload(LEDStrip *led_strip) {
	#define CHUNK_SIZE 16

	uint8_t r_chunk[CHUNK_SIZE];
	uint8_t g_chunk[CHUNK_SIZE];
	uint8_t b_chunk[CHUNK_SIZE];
	int i, k;

	for(i = 0; i < LED_ROWS*LED_COLS; i += CHUNK_SIZE) {
		for(k = 0; k < CHUNK_SIZE && i + k < LED_ROWS*LED_COLS; ++k) {
			r_chunk[k] = leds[i + k][R_INDEX];
			g_chunk[k] = leds[i + k][G_INDEX];
			b_chunk[k] = leds[i + k][B_INDEX];
		}

		if (led_strip_set_rgb_values(led_strip, i, k, r_chunk, g_chunk, b_chunk) < 0) {
			break;
		}
	}
}

void frame_prepare_next(void) {
	int offset = (int)rainbow_position % (LED_ROWS*LED_COLS);
	int count = LED_ROWS*LED_COLS - offset;

	memcpy(leds, &rainbow[offset], count * 3);
	memcpy(&leds[count], rainbow, offset * 3);

	rainbow_position += (LED_ROWS*LED_COLS) * RAINBOW_SPEED / 4.0;
}

// Frame rendered callback, is called when a new frame was rendered
void cb_frame_rendered(uint16_t length, void *user_data) {
	(void)length; // avoid unused parameter warning

	LEDStrip *led_strip = (LEDStrip *)user_data;

	frame_upload(led_strip);
	frame_prepare_next();
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	LEDStrip led_strip;
	led_strip_create(&led_strip, UID_LED_STRIP_BRICKLET, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		exit(1);
	}
	// Don't use device before ipcon is connected

	// Call a getter to check that the Bricklet is avialable
	uint16_t frame_duration;
	if(led_strip_get_frame_duration(&led_strip, &frame_duration) < 0) {
		fprintf(stderr, "Not Found: LED Strip (%s)\n", UID_LED_STRIP_BRICKLET);
		exit(1);
	}

	printf("Found: LED Strip (%s)\n", UID_LED_STRIP_BRICKLET);

	// Set frame duration to 20ms (50 frames per second)
	led_strip_set_frame_duration(&led_strip, 1000 / RAINBOW_FRAME_RATE);

	// Register frame rendered callback to function cb_frame_rendered
	led_strip_register_callback(&led_strip,
	                            LED_STRIP_CALLBACK_FRAME_RENDERED,
	                            (void *)cb_frame_rendered,
	                            (void *)&led_strip);

	int i;
	for(i = 0; i < LED_ROWS*LED_COLS; ++i) {
		hsv_to_rgb(255 * i / (LED_ROWS*LED_COLS), 255, 25, &rainbow[i][0], &rainbow[i][1], &rainbow[i][2]);
	}

	// Set initial rgb values to get started
	cb_frame_rendered(0, (void *)&led_strip);

	printf("Press key to exit\n");
	getchar();
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
}

File: rainbow/python/rainbow.py
#!/usr/bin/env python
# -*- coding: utf-8 -*

from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_led_strip import LEDStrip

import colorsys
import math

import config

class Rainbow:
    leds = [x[:] for x in [(0, 0, 0)]*config.LED_ROWS*config.LED_COLS]
    rainbow = [x[:] for x in [(0, 0, 0)]*config.LED_ROWS*config.LED_COLS]
    rainbow_position = 0

    def __init__(self, ipcon):
        self.okay = False
        self.ipcon = ipcon

        if not config.UID_LED_STRIP_BRICKLET:
            print("Not Configured: LED Strip (required)")
            return

        self.led_strip = LEDStrip(config.UID_LED_STRIP_BRICKLET, self.ipcon)

        try:
            self.led_strip.get_frame_duration()
            print("Found: LED Strip ({0})").format(config.UID_LED_STRIP_BRICKLET)
        except:
            print("Not Found: LED Strip ({0})").format(config.UID_LED_STRIP_BRICKLET)
            return

        for i in range(config.LED_ROWS*config.LED_COLS):
            r, g, b = colorsys.hsv_to_rgb(1.0*i/(config.LED_ROWS*config.LED_COLS), 1, 0.1)
            self.rainbow[i] = (int(r*255), int(g*255), int(b*255))

        self.okay = True

        self.update_frame_rate()
        self.led_strip.register_callback(self.led_strip.CALLBACK_FRAME_RENDERED,
                                         self.frame_rendered)

    def stop_rendering(self):
        if not self.okay:
            return

        self.led_strip.register_callback(self.led_strip.CALLBACK_FRAME_RENDERED,
                                         None)

    def update_frame_rate(self):
        if not self.okay:
            return

        self.led_strip.set_frame_duration(1000.0 / config.RAINBOW_FRAME_RATE)

    def frame_rendered(self, _):
        self.frame_upload()
        self.frame_prepare_next()

    def frame_upload(self):
        if not self.okay:
            return

        # Reorder LED data into R, G and B channel
        r = []
        g = []
        b = []
        for i in range(config.LED_ROWS*config.LED_COLS):
            r.append(self.leds[i][config.R_INDEX])
            g.append(self.leds[i][config.G_INDEX])
            b.append(self.leds[i][config.B_INDEX])

        # Make chunks of size 16
        r_chunk = [r[i:i+16] for i in range(0, len(r), 16)]
        g_chunk = [g[i:i+16] for i in range(0, len(g), 16)]
        b_chunk = [b[i:i+16] for i in range(0, len(b), 16)]

        for i in range(len(r_chunk)):
            length = len(r_chunk[i])

            # Fill up chunks with zeros
            r_chunk[i].extend([0]*(16-len(r_chunk[i])))
            g_chunk[i].extend([0]*(16-len(g_chunk[i])))
            b_chunk[i].extend([0]*(16-len(b_chunk[i])))

            try:
                self.led_strip.set_rgb_values(i*16, length, r_chunk[i], g_chunk[i], b_chunk[i])
            except:
                break

    def frame_prepare_next(self):
        self.leds = self.rainbow[int(self.rainbow_position) % (config.LED_ROWS*config.LED_COLS):] + self.rainbow[:int(self.rainbow_position) % (config.LED_ROWS*config.LED_COLS)]
        self.rainbow_position += (config.LED_ROWS*config.LED_COLS) * config.RAINBOW_SPEED / 4.0


if __name__ == "__main__":
    # Create IP Connection and connect it
    ipcon = IPConnection()
    ipcon.connect(config.HOST, config.PORT)

    # Create Rainbow object and start rendering
    rainbow = Rainbow(ipcon)
    rainbow.frame_rendered(0)

    raw_input('Press enter to exit\n') # Use input() in Python 3

    ipcon.disconnect()
